src/main/scala/mesosphere/marathon/api/forwarder/ServletOutputStreamSink.scala

```diff
@@ -0,0 +1,129 @@
+package mesosphere.marathon
+package api.forwarder
+
+import akka.stream.scaladsl.Sink
+import akka.stream.stage.{GraphStageLogic, InHandler}
+import akka.stream.{Attributes, Inlet}
+import akka.stream.stage.GraphStageWithMaterializedValue
+import akka.Done
+import akka.stream.SinkShape
+import akka.util.ByteString
+import com.typesafe.scalalogging.StrictLogging
+import java.util.concurrent.atomic.AtomicBoolean
+import javax.servlet.{AsyncContext, WriteListener}
+import scala.concurrent.{Future, Promise}
+import scala.util.Try
+
+/**
+  * Graph stage which implements a non-blocking io ServletOutupStream writer, following the protocol outlined here:
+  *
+  * http://www.oracle.com/webfolder/technetwork/tutorials/obe/java/HTML5andServlet31/HTML5andServlet%203.1.html#section3
+  *
+  * The following runtime restrictions apply:
+  *
+  * - This Sink cannot be materialized twice
+  * - No other writers for this outputStream may exist (IE no other component may register a writeListener)
+  * - The associated context must be put in to async mode, first
+  *
+  * Materialized Future will fail if the outputStream is not upgraded to async, or if a writeListener is already
+  * registered.
+  *
+  * Associated AsyncContext is not automatically closed; it is the responsible module using this sink to call
+  * asyncContext.complete() AFTER the materialized future is completed (failure or success).
+  */
+class ServletOutputStreamSink(asyncContext: AsyncContext) extends GraphStageWithMaterializedValue[SinkShape[ByteString], Future[Done]] with StrictLogging {
```

Comment:
So, we cannot use `asJavaStrem` because it's not a simple `OutputStream` but a `ServletOutputStream`?

Reply:
I don't understand, but yes, there are fundamental differences between OutputStream and ServletOutputStream.